8253064: monitor list simplifications and getting rid of TSM

dcubed-ojdk opened the pull request:
Changes from @fisk and @dcubed-ojdk to:

- simplify ObjectMonitor list management
- get rid of Type-Stable Memory (TSM)

This change has been tested with Mach5 Tier[1-3],4,5,6,7,8; no new regressions.
Aurora Perf runs have also been done (DaCapo-h2, Quick Startup/Footprint,
SPECjbb2015-Tuned-G1, SPECjbb2015-Tuned-ParGC, Volano)
  - a few minor regressions (<= -0.24%)
  - Volano is 6.8% better

Eric C. has also done promotion perf runs on these bits and says "the results look fine".

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253064](https://bugs.openjdk.java.net/browse/JDK-8253064): monitor list simplifications and getting rid of TSM


### Reviewers
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to aec90b9a54b16ecd4d3fce92ad4a1c3940c749c9
 * [Robbin Ehn](https://openjdk.java.net/census#rehn) (@robehn - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 61d36884fe4fc31b887cd7b6f3c090ea89d82adc


### Contributors
 * Erik Österlund `<eosterlund@openjdk.org>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/642/head:pull/642`
`$ git checkout pull/642`

dcubed-ojdk reviewed (commented):
Self review done.

dcubed-ojdk commented:
**Gory details about these changes from @fisk and @dcubed-ojdk:**

### Simplify `ObjectMonitor` List Management:

- delete per-thread in-use and free-lists.
- delete global free-list and global wait-list; there is a still a global in-use list; 
- move async deflation work from the `ServiceThread` to a dedicated `MonitorDeflationThread`; this prevents `ObjectMonitor` inflation storms from delaying the work done by the `ServiceThread` for other subsystems; this means the `ServiceThread` no longer wakes up every `GuaranteedSafepointInterval` to check for work.
- the `AllocationState` enum is dropped along with the  `_allocation_state` field and associates getters and setters; the simpler list management no longer requires the allocation state to be tracked.
- the safepoint cleanup phase no longer requests async monitor deflation; there is no longer a safepoint cleanup task for monitor deflation, but there is still an auditing/logging hook for debugging purposes.
- delete ObjectSynchronizer functions associated with more complicated list management: `deflate_global_idle_monitors()`, `deflate_per_thread_idle_monitors()`, `deflate_common_idle_monitors()`, `om_flush()`, `prepend_list_to_common()`, `prepend_list_to_global_free_list()`, `prepend_list_to_global_wait_list()`, `prepend_list_to_global_in_use_list()`, `prepend_to_common()`, `prepend_to_om_free_list()`, `prepend_to_om_in_use_list()`, `take_from_start_of_common()`, `take_from_start_of_global_free_list()`, `take_from_start_of_om_free_list()`
- delete the spin-lock functions needed by the more complicated list management.
- delete a number of audit/debug/logging related functions needed by the more complicated list management.
- restore the barrier related code that needed relocation due to om_flush()'s access of the weak obj reference; now that om_flush() is gone, the barrier related code can go back to its more natural place.

### Get Rid of Type-Stable Memory (TSM):

- `ObjectMonitor` now subclasses `CHeapObj<mtInternal>`.
- the `ObjectMonitor` constructor and destructor are now more normal C++!
- delete `ObjectMonitor` functions associated with TSM: `clear()`, `clear_common()`, `object_addr()`, `Recycle()`, and `set_object()`.
- delete the version of `set_owner_from()` that support two possible old values since it is no longer needed; we are not recycling deflated `ObjectMonitor`s anymore so there's not a possibility of a `NULL` `_owner` value or a `DEFLATER_MARKER` value on the same code path.
- delete ObjectSynchronizer functions associated with TSM: `om_alloc()`, `om_release()`, `prepend_block_to_lists()`
- simplify ObjectSynchronizer functions related to TSM: `deflate_idle_monitors()`, `deflate_monitor_list()`, `inflate()`

### Change A Displaced Header is Always at Offset 0

- Change `markWord::displaced_mark_helper()` and `markWord::set_displaced_mark_helper()` to no longer assume that the displaced header in a `BasicLock` or `ObjectMonitor` is at offset 0.
- ObjectMonitor::header_addr() no longer require the offset to be zero.

### New Diagnostic Options

- `AvgMonitorsPerThreadEstimate` - Used to estimate a variable ceiling based on number of threads for use with `MonitorUsedDeflationThreshold`; default is 1024, 0 is off, range is 0..max_jint. The current count of inflated `ObjectMonitor`s and the ceiling are used to determine whether the in-use ratio is higher than `MonitorUsedDeflationThreshold` (default 90).

- `MonitorDeflationMax` - The maximum number of `ObjectMonitor`s to deflate, unlink and delete at one time; default is 1 million; range is 1024..max_jint.

dcubed-ojdk commented:
/contributor add @fisk

dcubed-ojdk commented:
Rebased the project to jdk-16+23. Local macOS and Linux-X64 builds and
KitchensinkSanity runs pass. Kicking off a new round of Mach5 testing...

@coleenp, @dholmes-ora, @fisk, and @robehn - as usual for ObjectMonitor
stuff, your reviews would be greatly appreciated.

dholmes-ora commented on src/hotspot/share/runtime/objectMonitor.hpp:
```diff
@@ -139,20 +139,12 @@ class ObjectMonitor {
-  // choices to share the cache line with _header.
+  // have busy multi-threaded access. _header and _object are set at initial
+  // inflation. The _object does not change, so it is a good choice to share
+  // its the cache line with _header.
```
Typo: its the

dcubed-ojdk replied:
Nice catch. Fixed.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -57,6 +56,106 @@
 #include "utilities/preserveException.hpp"
 
+void MonitorList::add(ObjectMonitor* m) {
+  for (;;) {
```
Style nit: a do/while loop would like nicer here.

dcubed-ojdk replied:
Fixed:
```
  do {
    ObjectMonitor* head = Atomic::load(&_head);
    m->set_next_om(head);
  } while (Atomic::cmpxchg(&_head, head, m) != head);
```

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -57,6 +56,106 @@
+  while (m != NULL) {
+    if (m->is_being_async_deflated()) {
+      // Find next live ObjectMonitor.
+      ObjectMonitor* next = m;
+      while (next != NULL && next->is_being_async_deflated()) {
```
Nit: This loop seems odd. Given we know m is_being_async_deflated, this should either be a do/while loop, or else we should initialize:

ObjectMonitor* next = m->next_om();

and dispense with the awkwardly named next_next.

dcubed-ojdk replied:
@fisk - I'm leaving this one for you for now.

fisk replied:
Changing it to a do/while loop makes sense. The while condition is always true the first iteration, so doing a while or do/while loop is equivalent. If you find the do/while loop easier to read, then that sounds good to me.

dcubed-ojdk replied:
Okay. I've changed it to a do-while loop.

dholmes-ora replied:
The loop that was being discussed here is the one on line 93 of the original changeset:
`while (next != NULL && next->is_being_async_deflated()) {`
I made no comment on the outer while loop.

dcubed-ojdk replied:
You're correct. I got confused by the line renumbering due to other changes.
Fortunately, the outer while loop can also correctly and easily be changed
into a do-while loop so I'll keep that change. I'm looking at the correct inner
while loop now.

dcubed-ojdk replied:
The inner while loop is now converted into a do-while loop.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -57,6 +56,106 @@
+      m = m->next_om();
+    }
+
+    if (self->is_Java_thread() &&
```
Non-JavaThreads may have a monitor list ???

dcubed-ojdk replied:
This function, MonitorList::unlink_deflated(), may be called by
either the MonitorDeflationThread or the VMThread. The VMThread
does not need to block for the safepoint which is what the
`if (self->is_Java_thread()` prevents.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -57,6 +56,106 @@
+                     unlinked_count, ObjectSynchronizer::in_use_list_ceiling(), count(), max());
+      }
+
+      // Honor block request.
+      ThreadBlockInVM tbivm(self->as_Java_thread());
```
ThreadBlockInVM is generally used to wrap blocking code, not to cause the current thread to block (which it will do as a side-effect if a safepoint/handshake is requested). Surely this should just be call to `process_if_requested` (or the new `process_if_requested_with_exit_check`)?

dcubed-ojdk replied:
This kind of use of ThreadBlockInVM predates this changeset so while
the location is new, then code style is old, very old... I'll hold off changing
this for now.

coleenp replied:
I'd rather see ThreadBlockInVM as the convention of allowing the thread to block if a safepoint is requested.  The calls like process_if_requested are becoming alphabet soup and keep changing, so having TBIVM is better in my opinion.
That said, this is a strange usage.  This code appears three times.  It should be a function like allow_safepoint_block(LogStream* ls, timer), with some comment above.  Then it's clear that it's checking for a safepoint in a loop that could take a long time and the logging is ancillary.

dcubed-ojdk replied:
I previously looked at refactoring the three locations where
`ThreadBlockInVM` is used. The problem with the refactoring
is that the log messages and the parameters have some
differences and some commonalities. Each of these logging
sites is trying to communicate some local context that is
unique to that call site along with some global context that
might have changed from call site to call site.

I'll take another look at refactoring shortly and will let you
know what I come up with.

dcubed-ojdk replied:
I refactored common ThreadBlockInVM code block into ObjectSynchronizer::chk_for_block_req().
Please let me know if you're okay with it.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2030,296 +1406,42 @@ ObjectMonitor* ObjectSynchronizer::inflate(Thread* self, oop object,
-      if (mid->try_set_owner_from(DEFLATER_MARKER, NULL) != DEFLATER_MARKER) {
-        // Deferred decrement for the JT EnterI() that cancelled the async deflation.
-        mid->add_to_contentions(-1);
+    if (self->is_Java_thread() &&
```
Again unclear how a non-JavaThread is doing this? Isn't this always done by the MonitorDeflation thread??

dcubed-ojdk replied:
This function, ObjectSynchronizer::deflate_monitor_list(), may be called
by either the MonitorDeflationThread or the VMThread. The VMThread
does not need to block for the safepoint which is what the
if (self->is_Java_thread() prevents.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2030,296 +1406,42 @@ ObjectMonitor* ObjectSynchronizer::inflate(Thread* self, oop object,
-      mid = next;  // mid keeps non-NULL next's locked state
-      next = next_next;
+      // Honor block request.
+      ThreadBlockInVM tbivm(self->as_Java_thread());
```
Same comment as previous use of TBIVM. (It's hard to see in the PR UI how they two blocks relate.)

dcubed-ojdk replied:
This kind of use of ThreadBlockInVM predates this changeset so while
the location is new, then code style is old, very old... I'll hold off changing
this for now.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2333,192 +1455,108 @@ class HandshakeForDeflation : public HandshakeClosure {
-// This function is called by the ServiceThread to deflate monitors.
-// It is also called by do_final_audit_and_print_stats() by the VMThread.
-void ObjectSynchronizer::deflate_idle_monitors() {
+// This function is called by the MonitorDeflationThread to deflate
+// ObjectMonitors. It is also called via do_final_audit_and_print_stats()
+// by the VMThread.
```
Ah! I think this addresses my previous comments about a non-JavaThread doing this.

dcubed-ojdk replied:
Yup...

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2333,192 +1455,108 @@ class HandshakeForDeflation : public HandshakeClosure {
     if (self->is_Java_thread()) {
+      if (ls != NULL) {
+        timer.stop();
+        ls->print_cr("before handshaking: unlinked_count=" SIZE_FORMAT ", in_use_list stats: ceiling=" SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
```
Style nit: line too long

dcubed-ojdk replied:
Yeah... I keep going back and forth for long logging lines...
I reformatted that one and another one a few lines farther down.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2333,192 +1455,108 @@ class HandshakeForDeflation : public HandshakeClosure {
+    // After the handshake, safely free the ObjectMonitors that were
+    // deflated in this cycle.
+    size_t deleted_count = 0;
+    for (ObjectMonitor* monitor: delete_list) {
```
I didn't realize C++ has a "foreach" loop construct! Is this in our allowed C++ usage?

dcubed-ojdk replied:
@fisk - this one is for you... :-)

fisk replied:
Yeah this is one of the new cool features we can use. I thought it is allowed, because it is neither in the excluded nor undecided list of features in our doc/hotspot-style.md file.

dcubed-ojdk replied:
https://bugs.openjdk.java.net/browse/JDK-8254733

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2333,192 +1455,108 @@ class HandshakeForDeflation : public HandshakeClosure {
-      OM_PERFDATA_OP(Deflations, inc(local_deflated_count));
-    }
+        // Honor block request.
+        ThreadBlockInVM tbivm(self->as_Java_thread());
```
Ditto previous comments on use of TBIVM.

dcubed-ojdk replied:
This kind of use of ThreadBlockInVM predates this changeset so while
the location is new, then code style is old, very old... I'll hold off changing
this for now.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2681,407 +1707,102 @@ void ObjectSynchronizer::audit_and_print_stats(bool on_exit) {
-}
+// Check the in_use_list; log the results of the checks.
+void ObjectSynchronizer::chk_in_use_list(outputStream* out, int *error_cnt_p) {
+  size_t l_in_use_count = _in_use_list.count();
```
Style nit: what is this `l_` prefix? Is that a one or a  small L? why do we want want/need it? (Applies elsewhere too)

dcubed-ojdk replied:
The "l_" prefix is used for a local copy of a value where we want to
make sure that we use a consistent value for the check and for the
resulting audit/logging message.

This is not a new thing with this changeset and that style was used
in previous versions of the ObjectMonitor audit/logging code.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2681,407 +1707,102 @@ void ObjectSynchronizer::audit_and_print_stats(bool on_exit) {
+void ObjectSynchronizer::chk_in_use_entry(ObjectMonitor* n, outputStream* out,
+                                          int* error_cnt_p) {
+  if (n->owner_is_DEFLATER_MARKER()) {
+    // This should not happen, but if it does, it is not fatal.
```
Deflating an in-use monitor is not fatal? Please explain how things would recover.

dcubed-ojdk replied:
When the MonitorDeflationThread is doing its final async deflation cycle before
VM shutdown, it is possible for it to finish a deflation pass on the in-use list and
then block for the final safepoint before unlinking the deflated ObjectMonitors.

If the VMThread happens to be doing an audit_and_print_stats() call when this
happens (see the end of SafepointSynchronize::do_cleanup_tasks()), then we
don't want that audit to FAIL.

As for recovery, if the VMThread is doing a final audit, then when it sees an
already deflated ObjectMonitor on the in-use list, it will take care of the unlinking
and deleting...

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.hpp:
```diff
@@ -155,42 +160,21 @@ class ObjectSynchronizer : AllStatic {
-  // global list of blocks of monitors
-  static PaddedObjectMonitor* g_block_list;
+  static MonitorList   _in_use_list;
+  static jint          _in_use_list_ceiling;
```
Can you add some commentary on what this ceiling is as I could not understand its role just by looking at the code. Thanks.

dcubed-ojdk replied:
How about this:

dholmes-ora replied:
Thanks for the comment. So instead of checking the threshhold on each OM allocation we use this averaging technique to estimate the number of monitors in use? Can you explain how this came about rather than the simple/obvious check at allocation time. Thanks.

dcubed-ojdk replied:
I'm not sure I understand your question, but let me that a stab at it anyway...

We used to compare the sum of the in-use counts from all the in-use lists
with the total population of ObjectMonitors. If that ratio was higher than
MonitorUsedDeflationThreshold, then we would do an async deflation cycle.
Since we got rid of TSM, we no longer had a population of already allocated
ObjectMonitors, we had a max value instead. However, when the VMs use
of ObjectMonitors is first spinning up, the max value is typically very close
to the in-use count so we would always be asking for an async-deflation
during that spinning up phase.

I created the idea of a ceiling value that is tied to thread count and the
AvgMonitorsPerThreadEstimate to replace the population value that we
used to have. By comparing the in-use count against the ceiling value, we
no longer exceed the MonitorUsedDeflationThreshold when the VMs use
of ObjectMonitors is first spinning up so we no longer do async deflations
continuously during that phase. If the max value exceeds the ceiling value,
then we're using a LOT of ObjectMonitors and, in that case, we compare
the in-use count against the max to determine if we're exceeding the
MonitorUsedDeflationThreshold.

Does this help?

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -118,312 +217,13 @@ int dtrace_waited_probe(ObjectMonitor* monitor, Handle obj, Thread* thr) {
-// global list of blocks of monitors
-PaddedObjectMonitor* ObjectSynchronizer::g_block_list = NULL;
+MonitorList ObjectSynchronizer::_in_use_list;
+// Start the ceiling with one thread:
```
This relates to me not understanding what this ceiling is (as commented elsewhere) but why does this say "start with one thread" when the value of AvgMonitorsPerThreadEstimate defaults to 1024 ??

dcubed-ojdk replied:
The estimate is that a single thread will generate at most 1024 inflated
ObjectMonitors on average.

I changed the comment like this:
```
// Start the ceiling with the estimate for one thread:
jint ObjectSynchronizer::_in_use_list_ceiling = AvgMonitorsPerThreadEstimate;
```
Does that help?

dholmes-ora reviewed (commented):
Hi Dan,

Overall this looks great. Comparing old and new code is complex but the new code on its own is generally much simpler/clearer (not all though :) ).

I have a few nits, comments and queries below.

Thanks,
David

dcubed-ojdk commented on src/hotspot/share/runtime/monitorDeflationThread.hpp:
```diff
@@ -0,0 +1,48 @@
+  static void initialize();
+
+  // Hide this thread from external view.
+  bool is_hidden_from_external_view() const      { return true; }
```
Fixed.

I copied that from the serviceThread.hpp file, but, since I haven't otherwise
touched that file in this changeset, I'm going to leave serviceThread.hpp alone.

dcubed-ojdk commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -363,7 +377,7 @@ bool ObjectMonitor::enter(TRAPS) {
   if (event.should_commit()) {
     event.set_monitorClass(object()->klass());
-    event.set_address((uintptr_t)object_addr());
+    event.set_address((uintptr_t)this);
```
I noticed that in my preliminary review of Erik's changes. He checked
with the JFR guys and they said it just needed to be an address and
does not have to refer to the Object.

@fisk - can you think of a comment we should add here?

fisk replied:
We could write something along the lines of "An address that is 'unique enough', such that events close in time and with the same address are likely (but not guaranteed) to belong to the same object". This uniqueness property has always been more of a heuristic thing than anything else, as deflation shuffles the addresses around. Taking the this pointer vs an offset into the this pointer does however serve the exact same purpose; there was never any correlation to the contents of the object field.

dcubed-ojdk replied:
Thanks @fisk! I've added a slightly edited version of the comment:
```
    // Set an address that is 'unique enough', such that events close in
    // time and with the same address are likely (but not guaranteed) to
    // belong to the same object.
```
@dholmes-ora - does this work for you?

dcubed-ojdk commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -1351,7 +1469,7 @@ static void post_monitor_wait_event(EventJavaMonitorWait* event,
   event->set_monitorClass(monitor->object()->klass());
   event->set_timeout(timeout);
-  event->set_address((uintptr_t)monitor->object_addr());
+  event->set_address((uintptr_t)monitor);
```
I noticed that in my preliminary review of Erik's changes. He checked
with the JFR guys and they said it just needed to be an address and
does not have to refer to the Object.

@fisk - can you think of a comment we should add here?

fisk replied:
I wrote one in the section above, hope it is useful.

dcubed-ojdk replied:
Thanks @fisk. I copied the same comment here since it is about
1000 lines away from the other comment.

dcubed-ojdk commented:
@dholmes-ora - Thanks for the review! Hmm... I'm not sure why the GitHub UI
send out my replies one-at-a-time. Perhaps I should have replied from the
"files" view instead of the main PR view?

dcubed-ojdk commented:
It is a deflated monitor that is still on the in-use list.

dcubed-ojdk commented:
I've made all the changes based on @fisk's replies to @dholmes-ora comments.
@dholmes-ora - I think I've addressed all of your comments. Please let me know
if I've missing something.

dcubed-ojdk commented:
@dholmes-ora  - I'll stick with ceiling for now. If you're satisfied with the changeset,
please mark this PR as approved.

@fisk - Please mark this PR as approved if you're happy with the current version.

@robehn and @coleenp - It would be good to hear from one or both of you...

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -86,7 +86,7 @@ size_t MonitorList::unlink_deflated(Thread* self, LogStream* ls,
   ObjectMonitor* head = Atomic::load_acquire(&_head);
   ObjectMonitor* m = head;
-  while (m != NULL) {
+  do {
```
This wasn't the loop I was referring to. It is the while loop below this at line 93.

dcubed-ojdk replied:
The inner while loop is now converted into a do-while loop.

dcubed-ojdk commented:
Thanks for closing the loop on this one. Yes, next_next is needed
and it's a carry-over name from the baseline code.

And again, thanks for the thorough review.

fisk reviewed (approved):
Looks very good! Thanks for picking this up and taking it all the way!

dcubed-ojdk commented:
@fisk - Thanks for the sanity check review. And thanks for prototyping
this work and showing that this crazy idea could work! :-)

dcubed-ojdk commented:
@coleenp - I refactored common ThreadBlockInVM code block into ObjectSynchronizer::chk_for_block_req().
Please let me know if you're okay with it.

robehn commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2030,296 +1393,56 @@ ObjectMonitor* ObjectSynchronizer::inflate(Thread* self, oop object,
+    ls->print_cr("resuming %s: in_use_list stats: ceiling=" SIZE_FORMAT
+                 ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT, op_name,
+                 in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
+    timer_p->start();
```
ThreadBlockInVM have a miss-feature: it safepoint polls on front-edge and back-edge.
It should only have that poll on backedge, once that is fixed, this will do the wrong thing.
Also you may safepoint on both front-edge and back-edge now, so the timer would show the wrong thing then.

So to get the timer correct you should use:
SafepointMechanism::process_if_requested(thread);

dcubed-ojdk replied:
The baseline code (ObjectSynchronizer::deflate_common_idle_monitors())
uses ThreadBlockInVM currently. I don't want to change that as part of
this work. If we want to generally change uses of ThreadBlockInVM to
something else, then we should do that with a dedicated bug.

robehn replied:
Currently there is no issue with ThreadBlockInVM since there is no code inside those scopes.
This adds code there which assumes the timer will be 'resumed', and logs "resume" when it actually could be going to a safepoint.

dcubed-ojdk replied:
So if I narrow the scope around the ThreadBlockInVM, then it would be fine?

```
{
  // Honor block request.
  ThreadBlockInVM tbivm(self);
}
```

robehn replied:
Yes that avoids it!

dcubed-ojdk replied:
Done. I also did the one in ObjectSynchronizer::request_deflate_idle_monitors().

coleenp replied:
I like it but can you do one thing for me?  Can you s/chk/check/ in the name?

dcubed-ojdk replied:
I'd rather not. It is a function with six freaking parameters so
I went with names as short as I could...

robehn commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2333,192 +1456,100 @@ class HandshakeForDeflation : public HandshakeClosure {
+      if (self->is_Java_thread()) {
+        // A JavaThread must check for a safepoint/handshake and honor it.
+        chk_for_block_req(self->as_Java_thread(), "deletion", "deleted_count",
+                          deleted_count, ls, &timer);
```
If you release oopStorage when deflating you can do this entire loop while blocked instead, which will be faster.

dcubed-ojdk replied:
@fisk said we should not release the oopStorage during a safepoint
because that's not safe or will not be safe. I can't remember which.

robehn replied:
Yes that's why I said you can release it during deflation instead.

dcubed-ojdk replied:
What does this "you can do this entire loop while blocked instead" mean?

Releasing during deflation kind of messes with the life-cycle I was trying
to enforce since deletion is the nature end-of-life for these... But to think
about that I need to know what you mean by "you can do this entire loop
while blocked instead"...

robehn replied:
If you only need to free CHeap memory, you can do:
```
   size_t deleted_count = 0;
   ThreadBlockInVM tbivm(self);
    for (ObjectMonitor* monitor: delete_list) {
      delete monitor;
      deleted_count++;
    }
  }
```

dcubed-ojdk replied:
Ahhh... but that only works if we release the oopStorage when
we deflate. Okay. I grok it now, but don't want to do that in this
changeset. I would want a complete stress test cycle for that
kind of a change.

dcubed-ojdk replied:
Filed this new RFE:

robehn commented on src/hotspot/share/runtime/objectMonitor.hpp:
```diff
@@ -139,20 +139,12 @@ class ObjectMonitor {
   DEFINE_PAD_MINUS_SIZE(0, OM_CACHE_LINE_SIZE, sizeof(volatile markWord) +
-                        sizeof(WeakHandle) + sizeof(AllocationState));
+                        sizeof(WeakHandle));
   // Used by async deflation as a marker in the _owner field:
```
I have test with and without padding, I saw no difference.

dcubed-ojdk replied:
We've removed enough padding with this work already. If we
want to do more padding removal, then we need to use a
different RFE.

robehn replied:
Sure, this was more a FYI.

dcubed-ojdk replied:
Filed this new RFE:
    JDK-8256303 revisit ObjectMonitor padding between fields
    https://bugs.openjdk.java.net/browse/JDK-8256303

robehn reviewed (approved):
Hi, thanks for fixing.

I had some comments nothing so approving.

dcubed-ojdk commented on src/hotspot/share/runtime/monitorDeflationThread.cpp:
```diff
@@ -0,0 +1,98 @@
+        // Wait until notified that there is some work to do.
+        // We wait for GuaranteedSafepointInterval so that
+        // is_async_deflation_needed() is checked at the same interval.
+        ml.wait(GuaranteedSafepointInterval);
```
We may still decide to do this fix (even though the _object field is now weak):

    JDK-8249638 Re-instate idle monitor deflation as part of System.gc()
    https://bugs.openjdk.java.net/browse/JDK-8249638

and if we do, then we'll still need the request mechanism.

robehn replied:
So why not use a local semaphore and wait with safepoint check instead?

dcubed-ojdk replied:
Sorry my preference is for Monitors instead of semaphores. Let's
take that discussion off this PR and you can explain why you dislike
the Monitor so much and think the local semaphore is the way to go.

robehn replied:
Yes

dcubed-ojdk replied:
Filed the following new RFE:
    JDK-8256241 replace MonitorDeflation_lock with a semaphore
    https://bugs.openjdk.java.net/browse/JDK-8256241

dcubed-ojdk commented:
@robehn - Thanks for the review!! And thanks for approving.

coleenp commented on src/hotspot/share/runtime/synchronizer.hpp:
```diff
@@ -25,6 +25,7 @@
 #ifndef SHARE_RUNTIME_SYNCHRONIZER_HPP
 #define SHARE_RUNTIME_SYNCHRONIZER_HPP
 
+#include "logging/logStream.hpp"
```
If you need to put MonitorList in the header file, use a forward declaration for LogStream instead of #including logstream.hpp.

dcubed-ojdk replied:
I can see if that will work.

coleenp commented on src/hotspot/share/runtime/objectMonitor.hpp:
```diff
@@ -179,7 +171,7 @@ class ObjectMonitor {
   jint  _contentions;               // Number of active contentions in enter(). It is used by is_busy()
```
Future RFE - can we replace jint with int32_t or even int or some C++ types.  We're trying not to have Java types leak into runtime code since this doesn't directly interface with Java.

dcubed-ojdk replied:
It can be a future RFE, but it won't be at the top of my list of
things to do. There may already be an RFE for that.

coleenp replied:
No, I assume it's not high priority.  I'll file an RFE because someday I want these to be cleaned up as a personal nit.

dcubed-ojdk replied:
Filed the following new RFE:
    JDK-8256306 ObjectMonitor::_contentions field should not be 'jint'
    https://bugs.openjdk.java.net/browse/JDK-8256306

dcubed-ojdk commented on src/hotspot/share/oops/markWord.cpp:
```diff
@@ -25,9 +25,42 @@
+  // This should never happen:
+  fatal("bad header=" INTPTR_FORMAT, value());
+}
+
```
Yup. That was the point. Get rid of the magic offset zero coding...

dcubed-ojdk commented on src/hotspot/share/runtime/globals.hpp:
```diff
@@ -731,6 +731,20 @@ const intx ObjectAlignmentInBytes = 8;
+                                                                            \
   product(intx, MonitorUsedDeflationThreshold, 90, EXPERIMENTAL,            \
           "Percentage of used monitors before triggering deflation (0 is "  \
           "off). The check is performed on GuaranteedSafepointInterval "    \
```
Robbin added MonitorUsedDeflationThreshold as an experimental
option back in JDK10. See the longer reply to David's comment.
I don't plan to change that option with this changeset.

dcubed-ojdk replied:
Filed the following new RFE:
    JDK-8256304 should MonitorUsedDeflationThreshold be experimental or diagnostic
    https://bugs.openjdk.java.net/browse/JDK-8256304

dcubed-ojdk commented on src/hotspot/share/runtime/monitorDeflationThread.cpp:
```diff
@@ -0,0 +1,98 @@
+      // suspend-equivalent because MonitorDeflationThread is not
+      // visible to external suspension.
+
+      ThreadBlockInVM tbivm(jt);
```
Ummm... we use a JavaThread because we have to stop async deflation
during safepoints so that we're not messing with Object headers during GC.
Yes, it's possible to use a non-JavaThread because this is "just software",
but I don't want to try to figure out those races...

coleenp replied:
Ok, I see. This would be bad if it was not a JavaThread.

coleenp replied:
Ok, I see why it has to be a JavaThread.

dcubed-ojdk commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -2596,6 +1637,29 @@ u_char* ObjectSynchronizer::get_gvars_stw_random_addr() {
 }
 
+// Do the final audit and print of ObjectMonitor stats; must be done
+// by the VMThread (at VM exit time).
```
Okay.

dcubed-ojdk commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -475,6 +492,110 @@ int ObjectMonitor::TryLock(Thread * Self) {
+// Contending threads that see that condition know to retry their operation.
+//
+bool ObjectMonitor::deflate_monitor() {
+  if (is_busy()) {
```
Nice catch! That has been there for many, many years...

dcubed-ojdk replied:
Filed the following new RFE:
    JDK-8256301 ObjectMonitor::is_busy() should return bool
    https://bugs.openjdk.java.net/browse/JDK-8256301

dcubed-ojdk commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -475,6 +492,110 @@ int ObjectMonitor::TryLock(Thread * Self) {
+    // deflation dance.
+    if (try_set_owner_from(NULL, DEFLATER_MARKER) != NULL) {
+      // The owner field is no longer NULL so we lost the race since the
+      // ObjectMonitor is now busy.
```
No we can't assert that (contentions > 0). The ownership
might have been taken by a fast path thread so it grabbed
ownership without having to update contentions and wait.

dcubed-ojdk commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -475,6 +492,110 @@ int ObjectMonitor::TryLock(Thread * Self) {
+      // NULL if it is still DEFLATER_MARKER.
+      if (try_set_owner_from(DEFLATER_MARKER, NULL) != DEFLATER_MARKER) {
+        // Deferred decrement for the JT EnterI() that cancelled the async deflation.
+        add_to_contentions(-1);
```
No it is not a ref_count. We got rid of the accurate ref_count field
because maintaining it was too slow.

contentions just tells you how many threads have blocked on the
slow-path trying to enter the monitor. The fast-paths never touch
contentions. The comment on line 360 is accurate.

coleenp replied:
But contentions is used for more than informing JVMTI, it's used to test whether the monitor is_busy on the slow path.  That's why I wanted the comment to say something like your last sentence, since I spent time trying to understand why the various calls to add_to_contentions(-1) in deflate_monitor earlier.

dcubed-ojdk replied:
Ahhh... I think I understand your confusion. This line:

L550: // Deferred decrement for the JT EnterI() that cancelled the async deflation.
L551:  add_to_contentions(-1);

doesn't match up with this line:

L361: add_to_contentions(1);

It matches up with one of these:
```
  if (try_set_owner_from(DEFLATER_MARKER, Self) == DEFLATER_MARKER) {
    // Cancelled the in-progress async deflation by changing owner from
    // DEFLATER_MARKER to Self. As part of the contended enter protocol,
    // contentions was incremented to a positive value before EnterI()
    // was called and that prevents the deflater thread from winning the
    // last part of the 2-part async deflation protocol. After EnterI()
    // returns to enter(), contentions is decremented because the caller
    // now owns the monitor. We bump contentions an extra time here to
    // prevent the deflater thread from winning the last part of the
    // 2-part async deflation protocol after the regular decrement
    // occurs in enter(). The deflater thread will decrement contentions
    // after it recognizes that the async deflation was cancelled.
    add_to_contentions(1);
```

The long comments are in the two places where we temporarily
increment contentions to stop the race with the deflater thread
and the shorter comment, e.g., L550, are for where we undo the
temporary increment.

The primary purpose of the contentions field is for JVM/TI and
M&M queries. We just (temporarily) steal it for async deflation
purposes...

coleenp replied:
Well since it controls async deflation, it should probably get a mention since this comment on its own is not true:

// Keep track of contention for JVM/TI and M&M queries and control async deflation.

dcubed-ojdk replied:
The comment for **that** location is correct. That's the location
where we keep track of contended monitors for JVM/TI and M&M.
The comments at the other locations are also correct (very verbose
for two of them, but correct).

dcubed-ojdk replied:
Look at the entire context of that location:

dcubed-ojdk commented on src/hotspot/share/runtime/synchronizer.hpp:
```diff
@@ -34,16 +35,34 @@
+  Iterator(ObjectMonitor* head) : _current(head) {}
+  bool has_next() const { return _current != NULL; }
+  ObjectMonitor* next();
+};
```
I can see if that will work.

dcubed-ojdk commented:
@robehn and @coleenp - resolved more comments.

dholmes-ora commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -118,312 +230,24 @@ int dtrace_waited_probe(ObjectMonitor* monitor, Handle obj, Thread* thr) {
+// ObjectMonitors than the estimated average.
+//
+// Start the ceiling with the estimate for one thread:
+jint _in_use_list_ceiling = AvgMonitorsPerThreadEstimate;
```
Why is this a jint when you use size_t for its accessor and all the other sizes that you compare with the ceiling are also size_t?
I'm not sure size_t is right to use in these cases (do we really expect different maximums on 32-bit versus 64-bit?) but it should be all or none IMO.

coleenp replied:
Our int types are really confused.  AvgMonitorsPerThreadEstimate is defined as an intx which is intptr_t and the range of it is 0..max_jint which is 0 .. 0x7fffffff . jint is long on windows (the problematic type) and int on unix.  Since this is a new declaration, it probably should be something other than jint but what?
At any rate, it should be declared as 'static'.

dcubed-ojdk replied:
`_in_use_list_ceiling` is a jint **because** we've specified the range
as `0..max_jint` and I wanted some sanity to that variable's type.

If I change `_in_use_list_ceiling` to `size_t`, then I get a compile time
error probably because  `AvgMonitorsPerThreadEstimate` is an `intx`
which (I think) is my only choice for a command line option.

@fisk will have to chime in with the background on why he picked `size_t`.

dcubed-ojdk replied:
@coleenp - Nice catch on the missing 'static'.

fisk replied:
I typically use size_t for entities that can scale with the size of the machine's memory, so I don't have to think about whether there are enough bits. Could AvgMonitorsPerThreadEstimate be uintx instead of intx? And then maybe we don't need to declare a range, as the natural range of the uintx seems perfectly valid.

dcubed-ojdk replied:
I'm pretty sure I copied the decl for AvgMonitorsPerThreadEstimate
from some other already existing option. That's SOP for me anyway...
If we make any more changes here it will have to be in a follow up.

dcubed-ojdk replied:
Filed the following new RFE:
    JDK-8256307 cleanup AvgMonitorsPerThreadEstimate and _in_use_list_ceiling types
    https://bugs.openjdk.java.net/browse/JDK-8256307

dholmes-ora reviewed (changes requested):
One change requested in relation to use of jint instead of size_t.
One code simplification suggestion.
Thanks,
David

robehn commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -1381,7 +1220,9 @@ bool ObjectSynchronizer::request_deflate_idle_monitors() {
+      {
+        ThreadBlockInVM tbivm(self->as_Java_thread());
+      }
       os::naked_short_sleep(999);  // sleep for almost 1 second
```
Hi Dan, you need to be blocked while sleeping, otherwise you are blocking safepoints for "almost 1 second".
So previous code was correct.

dcubed-ojdk replied:
Good catch! Yes, I should not have written that change so late at
night. I know better. I have reverted it.

coleenp reviewed (changes requested):
Looks good!  I don't have more comments.

dcubed-ojdk commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -122,23 +148,11 @@ size_t MonitorList::unlink_deflated(Thread* self, LogStream* ls,
-      }
+    if (self->is_Java_thread()) {
+      // A JavaThread must check for a safepoint/handshake and honor it.
+      ObjectSynchronizer::chk_for_block_req(self->as_Java_thread(), "unlinking",
```
At this point, the refactoring has grown on me. I like the fact that
it reduces the "noise" in those three functions.

dcubed-ojdk commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -1191,7 +1220,9 @@ bool ObjectSynchronizer::request_deflate_idle_monitors() {
     }
     if (self->is_Java_thread()) {
       // JavaThread has to honor the blocking protocol.
-      ThreadBlockInVM tbivm(self->as_Java_thread());
+      {
+        ThreadBlockInVM tbivm(self->as_Java_thread());
+      }
       os::naked_short_sleep(999);  // sleep for almost 1 second
     } else {
       os::naked_short_sleep(999);  // sleep for almost 1 second
```
As @robehn pointed out, I screwed up that change because
I lost the block over duration of the sleep. I've reverted that
change to the original.

dcubed-ojdk commented:
@dholmes-ora, @robehn and @coleenp - Thanks for chiming in on the
review again.

For the first time, I have a real conflict in a file so I'm updating my repo
to the latest and greatest to see how that works. Then I'll make the
minor tweaks that I mumbled about above.

I did a Mach5 Tier1,2,3,4,5,6,7 last night and got through it with a
single Tier7 unrelated and known failure. Of course, the baseline
for last night was jdk-16+23 so...

robehn reviewed (approved):
Thanks!

dcubed-ojdk commented:
Rebuild on my MBP13 and KitchensinkSanity look good.
/integrate

dcubed-ojdk commented:
This PR is now integrated! @dholmes-ora - for some reason you are not listed
as a reviewer and I'm not sure why. I'm sorry I didn't notice that you were dropped
off before I pulled the trigger.
